src/jdk.incubator.code/share/classes/jdk/incubator/code/dialect/core/CoreOp.java

```diff
@@ -318,6 +320,85 @@ public Block.Builder lower(Block.Builder b, CodeTransformer _ignore) {
+                        throw new RuntimeException(e);
+                    }
+                    if (invokeOpCalledMethod instanceof Method m) {
+                        CoreOp.FuncOp f = Op.ofMethod(m).orElse(null);
+                        if (f != null) {
+                            Op.Result result = blockBuilder.op(CoreOp.funcCall(
+                                    iop.invokeDescriptor().name(),
+                                    f.invokableType(),
+                                    blockBuilder.context().getValues(iop.operands())));
+                            blockBuilder.context().mapValue(op.result(), result);
+                            return blockBuilder;
+                        }
+                    }
+                }
+                blockBuilder.op(op);
+                return blockBuilder;
+            });
+        }
+
+        static CoreOp.FuncOp invokeToFuncOp(JavaOp.InvokeOp invokeOp, MethodHandles.Lookup l) {
+            try {
+                Method method = invokeOp.invokeDescriptor().resolveToMethod(l);
+                return Op.ofMethod(method).orElse(null);
+            } catch (ReflectiveOperationException e) {
+                throw new IllegalStateException("Could not resolve invokeOp to method");
+            }
+        }
+
+        /**
+         * Returns a moduleOp with the given funcOp as the root.
+         * The funcOps in the moduleOp functionTable are returned in the order encountered within the funcOp.
+         */
+        public static CoreOp.ModuleOp ofFuncOp(CoreOp.FuncOp f, MethodHandles.Lookup l) {
+            Set<CoreOp.FuncOp> visited = new HashSet<>();
+            List<CoreOp.FuncOp> funcs = new ArrayList<>();
+            Deque<CoreOp.FuncOp> stack = new LinkedList<>();
+
+            stack.add(f);
+            while (!stack.isEmpty()) {
+                CoreOp.FuncOp cur = stack.removeLast();
+                if (!visited.add(cur)) continue;
+                funcs.add(convertInvokeToFuncCallOp(cur, l));
+                List<CoreOp.FuncOp> temp = new LinkedList<>();
+                cur.elements().filter(e -> e instanceof JavaOp.InvokeOp).forEach(e -> {
+                    if (invokeToFuncOp((JavaOp.InvokeOp) e, l) instanceof CoreOp.FuncOp resolvedFuncOp) {
+                        temp.addFirst(resolvedFuncOp);
+                    }
+                });
+                stack.addAll(temp);
+            }
+            return CoreOp.module(funcs);
+        }
+
+        /**
+         * Returns a moduleOp with a root funcOp representing the given lambdaOp.
+         * The funcOps in the moduleOp functionTable are returned in the order encountered within the lambdaOp.
+         */
+        public static CoreOp.ModuleOp ofLambdaOp(JavaOp.LambdaOp lambdaOp, MethodHandles.Lookup l, String lambdaName) {
+            boolean methodNameExists = Arrays.stream(l.lookupClass().getMethods())
+                    .anyMatch(method -> method.getName().equals(lambdaName));
```

Comment:
I don't think that check is right. To do this properly we would have to check against the names of methods encountered in the call graph. In general we have another naming clash problem with overloaded methods. A simple approach is to always postfix the function name with the topologically sorted index e.g. "<name>_<index>"